fix idempotency in kubernetes state module

djivey opened the pull request:
### What does this PR do?

### What issues does this PR fix or reference?
Fixes: This PR addresses idempotency issues in the saltext-kubernetes extension by implementing proper patch functionality for Kubernetes resources. Previously, the extension lacked the ability to intelligently update existing resources, leading to unnecessary recreation operations that could disrupt running workloads. This enhancement enables Salt to detect differences between desired and current state, applying only the necessary changes through Kubernetes patch operations, ensuring idempotent resource management.

closes #16 

### Previous Behavior
The state module was not idempotent for most present and absent functions.

### New Behavior
This ensures that running `salt '*' state.apply` multiple times produces consistent results without unintended side effects, which is fundamental to Salt's declarative infrastructure management approach.

- Added `patch_<resource>()` functions to support both dictionary patches and source file rendering
- Functions that previously returned `.to_dict()` responses now return `sanitize_for_serialization()` responses. The structure is similar but may have slight differences in nested object handling.
- Resources are now patched rather than replaced when differences are detected
- State operations correctly report "no changes" when resources already match desired state
- Eliminated unnecessary resource recreation that could cause service disruption
- Enhanced state logic to properly handle both creation and modification scenarios
- Added replace as an argument incase someone wanted to force a replacement of a resource
- Added dry_run to help with determining changes
- Updated test to reflect changes

### Merge requirements satisfied?
**[NOTICE] Bug fixes or new features require tests.**
<!-- Please review the [salt-extensions](https://salt-extensions.github.io/salt-extension-copier/topics/testing/writing.html) and [Salt test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests for your changes. -->
- [x ] Docs
- [ x] Changelog - https://salt-extensions.github.io/salt-extension-copier/topics/documenting/changelog.html#procedure
- [x ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

djivey committed:
fix deployment idempotency and add patching to deployments

djivey commented:
@ jeanluc Can you review the approach I used here to achieve idempotency? I only made changes to one resource type to keep the changes as low as I can for you. 

This is going to end up being another big change I believe, but if you can agree with the approach, it should be easy for me to apply across the rest of the resource types.

I am not sure that some of these tests like the integration tests are needed, but I wanted to at least show the changes worked.

lkubb commented:
Afaict it's just https://github.com/salt-extensions/salt-extension-copier/issues/199. Sorry, forgot about this issue, will try to get a new template release out soon.

lkubb commented on changelog/16.changed.md:
```diff
@@ -0,0 +1 @@
+Functions that previously returned `.to_dict()` responses now return `sanitize_for_serialization()` responses. The structure is similar but may have slight differences in nested object handling.
```
suggestion: I don't think we should expect users to know about Kubernetes Python API client internals. Maybe phrase it like this:

```suggestion
API response transformation has been altered. All functions that returned resource dictionaries with `snake_case` keys (such as `show_deployment`) now return keys in `camelCase` instead, as found in YAML manifest files. The structure is similar, but may have slight differences in nested object handling. Dates and times might be reported differently.
```

lkubb commented on src/saltext/kubernetes/states/kubernetes.py:
```diff
@@ -344,9 +372,80 @@ def deployment_present(
+            if __opts__["test"]:
+                ret["result"] = None
+                ret["comment"] = "The deployment is going to be patched"
+                ret["changes"] = {"old": deployment, "new": res}
```
tip: At least when altering resources, I'd consider using Salt's RecursiveDictDiffer to only report changed attributes since the old/new format can become pretty verbose for large dicts:

```py
from salt.utils.dictdiffer import RecursiveDictDiffer

# ...

diff = RecursiveDictDiffer(deployment, res)
# exact implementation depends on the format of both dicts,
# if all keys are always present you can rely on .changed()
# only and create the differ with ignore_missing_keys=True
ret["changes"] = {
    "changed": diff.changed()
    "added": diff.added()
    "removed": diff.removed()
}

# or

ret["changes"] = diff.changed()
```

When creating fresh resources, I'd probably just do `ret["changes"]["created"] = name`, but that's definitely up to personal taste.

lkubb commented on src/saltext/kubernetes/states/kubernetes.py:
```diff
@@ -344,9 +372,80 @@ def deployment_present(
+                    **kwargs,
+                )
+
+            if res == deployment:
```
hint: I'd double-check that we can expect this simple check to be reliable since each dict is the result of a different API call (slight differences in return formats between API functions, missing/additional keys or time-dependent values).

lkubb commented on tests/functional/states/test_kubernetes.py:
```diff
@@ -363,37 +363,33 @@ def test_deployment_present(kubernetes, deployment, testmode, kubernetes_exe):
-#     assert ret.result is True
-#     assert "already exists" in ret.comment
-#     assert not ret.changes
+def test_deployment_present_idempotency(kubernetes, deployment):
+    """
+    Test kubernetes.deployment_present is idempotent
+    """
+    ret = kubernetes.deployment_present(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        spec=deployment["spec"],
+        wait=True,
+    )
```
nit: I'd also test with `test=true` for completeness' sake.

```suggestion
def test_deployment_present_idempotency(kubernetes, deployment, testmode):
    """
    Test kubernetes.deployment_present is idempotent
    """
    ret = kubernetes.deployment_present(
        name=deployment["name"],
        namespace=deployment["namespace"],
        spec=deployment["spec"],
        wait=True,
        test=testmode,
    )
```

lkubb commented on tests/functional/states/test_kubernetes.py:
```diff
@@ -420,6 +416,65 @@ def test_deployment_present_replace(kubernetes, deployment, kubernetes_exe):
     assert deployment_state["spec"]["replicas"] == 3
 
 
+def test_deployment_present_patch(kubernetes, deployment, kubernetes_exe):
```
suggestion: I'd test in test mode as well and assert that the resource is not changed. It's often neglected when writing a state, but can have surprising consequences if not handled correctly.

lkubb commented on tests/functional/states/test_kubernetes.py:
```diff
@@ -420,6 +416,65 @@ def test_deployment_present_replace(kubernetes, deployment, kubernetes_exe):
+    assert deployment_state["spec"]["replicas"] == 4
+
+
+def test_deployment_present_patch_source(
```
suggestion: Test mode, see above

lkubb commented on tests/unit/modules/test_kubernetesmod.py:
```diff
@@ -133,37 +142,49 @@ def test_delete_deployments(mock_kubernetes_lib):
+            )
+
+            result = kubernetes.delete_deployment("test")
+            assert result == {"code": 200}
+            assert kubernetes.kubernetes.client.AppsV1Api().delete_namespaced_deployment.called
```
note: This test is an example for my remark on unit tests above. It only checks that `delete_deployment` returns the value we made the `sanitize_for_serialization` mock return and ensures it calls `delete_namespaced_deployment`. For all this brittle setup, it's quite isolated from reality and redundant since we have the functional tests.

Both assertions should be covered by the functional test. The unit test however does not check for example that

* the deployment was actually deleted
* the client lib still has the `delete_namespaced_deployment`/`sanitize_for_serialization` etc. functions
* the client lib still works with the Kubernetes API
* etc.

Unit tests, especially with comprehensive client libs like the kubernetes one, often need a lot of surgical patching (which can make changes to the functions under test quite annoying and limits their meaningfulness).

I'd reserve unit tests for hard to test behavior like some unhappy paths, e.g. "Does the state module handle situations where an execution module raises a `CommandExecutionError`".

lkubb commented on tests/unit/modules/test_kubernetesmod.py:
note: Imo many of the current unit tests don't make sense since they test the same happy-path behavior as the functional (and integration) ones. I'd scrap the existing redundant ones instead of rewriting them and would advise to focus on comprehensive functional ones instead.

[See related comment on `test_delete_deployments`]

lkubb commented on tests/unit/states/test_kubernetes.py:
suggestion: Same comment regarding unit tests here. I'd scrap most of the existing (redundant) ones instead of rewriting them/adding new ones.

lkubb reviewed (commented):
If you merge https://github.com/salt-extensions/saltext-kubernetes/pull/21 and rebase this branch, CI should pass. 👍 

This looks pretty good! I added some comments that hopefully help guide you for the rest of the resources.

Once again thanks for your efforts :)

djivey commented on src/saltext/kubernetes/states/kubernetes.py:
```diff
@@ -296,7 +306,29 @@ def deployment_present(
             ret["comment"] = "The deployment is going to be created"
+            # Simulate creation to show changes to account for using the source argument
+            try:
+                dry_run_result = __salt__["kubernetes.create_deployment"](
```
I am not sure I follow what you mean here exactly. The primary reason I added the dry_run here is so that I can get a change result if using source because at this point that file has not been rendered yet. Using dry_run gets that to render and I can return results on the changes. 

Before it was only checking and returning on metadata and spec.

djivey replied:
I think I am going to change my approach here looking at all this again. I am going to drop replace from the present functions - I don not believe that should be default behavior for a present function. Instead, if users want to replace a resource, they can run the execution module replace functions separately. I am also, removing the dry-run as comparison logic. Instead use dry-run in test=true to return changes, and apply patch if test=true is not set.  example below

```python
    if __opts__["test"]:
        try:
            res = __salt__["kubernetes.patch_deployment"](
                name, namespace, dry_run=True, **patch_kwargs, **kwargs,
            )
        except CommandExecutionError as err:
            ret["result"] = None
            ret["comment"] = (
                "The deployment is going to be updated. "
                f"Dry run failed, possibly due to dependencies not created yet: {err}"
            )
            return ret

        if res == deployment:
            ret["result"] = True
            ret["comment"] = "The deployment is already in the desired state"
            return ret

        ret["result"] = None
        ret["comment"] = "The deployment is going to be updated"
        ret["changes"] = {"old": deployment, "new": res}
        return ret
```

djivey commented:
My apologies to anyone waiting on this PR. I have been extremely busy, but I plan on getting back to this next week.

djivey committed:
fix idempotency and update tests

djivey commented:
@lkubb If you have moment, can you give this a quick glance? Mostly, just to validate the idempotency aspect of it. 

A couple of notes:
- I did not add patch support for namespaces and pods. 
  - Namespaces because there isn't really much to them
  - Pods because they are almost entirely immutable.
- I dropped replace from the present functions as I do not think that should be the default behavior.
- I changed test mode to use dry-run for the changes that are going to be used and reordered all of the logic.
- I dropped most of the redundant unit tests and added new ones.

Due to the amount of changes, I just wanted a second set of eyes to make sure I didn't miss anything.

Also, a couple questions:

1. I do not see a way to trigger that last ci dependency. `CI / Set the CI Pipeline Exit StatusExpected — Waiting for status to be reported`
2. When this merges, does the release bot automatically update the https://github.com/salt-extensions/saltext-kubernetes/pull/18 as I did not merge that because I wanted to have this as part of that version.

lkubb commented on src/saltext/kubernetes/states/kubernetes.py:
```diff
@@ -343,109 +326,77 @@ def deployment_present(
-    if deployment is not None:
-        try:
-            if source:
+        if __opts__["test"]:
```
nit: I'd recommend unifying both logics:

```
try:
    res = __salt__["kubernetes.patch_deployment"]( # ...
              dry_run=bool(__opts__["test"]),
          )
except CommandExecutionError as err:
    if not __opts__["test"]:
        raise
    # ...
```

The same for other state functions in here.

lkubb commented on tests/functional/modules/test_kubernetesmod.py:
```diff
@@ -219,6 +246,88 @@ def test_create_existing_secret(kubernetes, secret, secret_data):
+    assert res is None
+
+
+def test_replace_secret(kubernetes, secret, secret_data):
```
nit: Since `replace` and `patch` have different semantics, I'd test that `replace` removes unspecified secret keys while `patch` leaves them intact (IIUC).

Unsure if that's as easy to do for other resource types, if not, running the same test against both actions is completely fine.

lkubb commented on tests/functional/states/test_kubernetes.py:
```diff
@@ -420,18 +392,22 @@ def test_deployment_present_patch(kubernetes, deployment, kubernetes_exe):
+    if not testmode:
+        assert deployment_state["spec"]["replicas"] == 4
+    else:
+        assert deployment_state["spec"]["replicas"] != 4
```
suggestion: Just a hint, it's possible to do this in one line:

`assert (deployment_state["spec"]["replicas"] == 4) is not testmode`

lkubb commented on tests/functional/states/test_kubernetes.py:
```diff
@@ -1205,20 +1192,20 @@ def test_node_label_present(kubernetes, labeled_node, testmode, kubernetes_exe):
-        assert not ret.changes
+        assert ret.changes["new"][label_name] == label_value
+        assert label_name not in ret.changes["old"]
         assert "The label is going to be set" in ret.comment
         # Verify label is not created in test mode
         node_label_state = kubernetes_exe.node_labels(labeled_node["name"])
         assert label_name not in node_label_state
```
nit: I'd minimize special-casing of test mode handling to the parts that differ:

```
assert ret.changes["new"][label_name] == label_value
assert label_name not in ret.changes["old"]
node_label_state = kubernetes_exe.node_labels(labeled_node["name"])
if not testmode:
    assert node_label_state[label_name] == label_value
else:
    assert label_name not in node_label_state
    assert "The label is going to be set" in ret.comment
```

Note that you don't need to `assert ret.result is True`, it's implicit from the assertions above.

lkubb commented on tests/unit/modules/test_kubernetesmod.py:
```diff
@@ -63,128 +63,16 @@ def mock_kubernetes_lib():
-        assert kubernetes.kubernetes.client.AppsV1Api().create_namespaced_deployment.called
+    with patch("saltext.kubernetes.modules.kubernetesmod._setup_conn", return_value={}):
+        with patch("saltext.kubernetes.modules.kubernetesmod._cleanup"):
+            with patch("saltext.kubernetes.modules.kubernetesmod.kubernetes") as mock_k8s:
```
nit: I recommend always passing `autospec=True` when patching functions, which ensures passed arguments are validated:

```
    with patch("saltext.kubernetes.modules.kubernetesmod._setup_conn", return_value={}, autospec=True):
        with patch("saltext.kubernetes.modules.kubernetesmod._cleanup", autospec=True):
            with patch("saltext.kubernetes.modules.kubernetesmod.kubernetes", autospec=True) as mock_k8s:
```

lkubb commented on tests/unit/modules/test_kubernetesmod.py:
```diff
@@ -338,44 +226,61 @@ def test_service_validation(mock_kubernetes_lib, invalid_spec, expected_error):
-            )
+def test_deployment_invalid_spec(invalid_spec, expected_error):
+    """Test deployment spec validation raises appropriate errors"""
+    func = getattr(kubernetes, "__dict_to_deployment_spec")
+    with pytest.raises(CommandExecutionError, match=expected_error):
+        func(invalid_spec)
```
nit: Does `kubernetes.__dict_to_deployment_spec(invalid_spec)` not work? Same below.

djivey replied:
It works, making that change.

lkubb commented on tests/unit/modules/test_kubernetesmod.py:
```diff
@@ -387,53 +292,475 @@ def test_deployment_invalid_spec(mock_kubernetes_lib, invalid_spec, expected_err
+                )
+
+                with pytest.raises(CommandExecutionError):
+                    kubernetes.deployments()
```
suggestion: Use the mock fixture `mock_api` instead.

```
def test_deployments_handles_api_exception(mock_api):
    """
    Test that deployments() handles ApiException properly
    """
    # Simulate API exception (e.g., network error, auth failure)
    mock_api.client.AppsV1Api().list_namespaced_deployment.side_effect = ApiException(
        status=500, reason="Internal Server Error"
    )

    with pytest.raises(CommandExecutionError):
        kubernetes.deployments()
```

Same below.

lkubb commented on src/saltext/kubernetes/modules/kubernetesmod.py:
```diff
@@ -189,12 +191,15 @@ def nodes(**kwargs):
-        log.exception("Exception when calling CoreV1Api->list_node")
-        raise CommandExecutionError(exc)
+        log.error("Exception when calling CoreV1Api->list_node", exc_info_on_loglevel=logging.DEBUG)
+        raise CommandExecutionError(exc) from exc
```
nit: Unsure if we need to log the exception when re-raising from it since Salt logs exceptions the same way, which duplicates the log entry.

Same below.

lkubb commented on src/saltext/kubernetes/states/kubernetes.py:
```diff
@@ -343,109 +326,77 @@ def deployment_present(
-                return ret
+            ret["result"] = None
+            ret["comment"] = "The deployment is going to be updated"
+            ret["changes"] = {"old": deployment, "new": res}
```
suggestion: Consider using `salt.utils.dictdiffer.RecursiveDictDiffer` when large dicts are involved in change reporting.

lkubb reviewed (commented):
> * I do not see a way to trigger that last ci dependency. `CI / Set the CI Pipeline Exit StatusExpected — Waiting for status to be reported`

Sorry about that, that was a branch protection misconfiguration since the name of the required job changed during the last template update. I adjusted the config, re-running CI should fix it.

> * When this merges, does the release bot automatically update the [Release v2.0.0 #18](https://github.com/salt-extensions/saltext-kubernetes/pull/18) as I did not merge that because I wanted to have this as part of that version.

Yes, just wait for the main branch CI to finish after merging it, updating the release PR is part of the main branch `push` workflow.

---

This looks awesome! Just some minor nits that you can choose to ignore. :)

djivey commented:
> > * I do not see a way to trigger that last ci dependency. `CI / Set the CI Pipeline Exit StatusExpected — Waiting for status to be reported`
> 
> Sorry about that, that was a branch protection misconfiguration since the name of the required job changed during the last template update. I adjusted the config, re-running CI should fix it.
> 
> > * When this merges, does the release bot automatically update the [Release v2.0.0 #18](https://github.com/salt-extensions/saltext-kubernetes/pull/18) as I did not merge that because I wanted to have this as part of that version.
> 
> Yes, just wait for the main branch CI to finish after merging it, updating the release PR is part of the main branch `push` workflow.
> 
> This looks awesome! Just some minor nits that you can choose to ignore. :)

Thank you for your feedback, and my apologies for the long delay in continuing support here. I have been extremely busy.

djivey committed:
use RecursiveDictDiffer, mock_api, clean uplogging, use autospec, and unify logic
